Language test improvements

Ericson2314 opened the pull request:
# Motivation

* lang now verifies errors and parse output
* Some new miscellaneous tests
* Easy way to update the tests

# Context

~~Depends on #5754~~
~~Depends on #8571~~

Based on original PR #633. Thanks @Mathnerd314 for writing that. 

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

edolstra commented:
I'm not sure we should have exact tests for error messages. It makes improving error messages more painful, and we don't give any particular guarantees about the contents of error messages (or any human-readable messages on stderr).

The empty `eval-okay-*.err.exp` files seem like unnecessary clutter in the repo, since succeeding tests should not print error messages. It's probably better to check that if `<test>.err.exp` does not exist, then the test should not produce any output on stderr.

Mathnerd314 commented:
Well, the error messages can be regenerated if they change. The point is really to track the error messages as part of the committing process. Even if nobody looked at the messages as part of the review process, it would still be useful for regression purposes to be able to track down when a message changed.

Ericson2314 commented:
> The empty `eval-okay-*.err.exp` files seem like unnecessary clutter in the repo...

Agreed. Good idea. They are no gone. `doAccept` won't recreate them either, and in fact will remove such empty files that were committed by mistake.

Mathnerd314 commented:
I should also note that a significant portion of GHC's testsuite is based on these "golden" tests that check the expected output, that's where I got the idea: https://gitlab.haskell.org/ghc/ghc/-/tree/master/testsuite/tests/typecheck/should_fail

Ericson2314 commented:
As to the larger question, I agree with @Mathnerd314, and I think also @roberth was bringing this up.

Simply put, just because we don't promise something is stable doesn't mean we want to change it by accident. Regressions are still bad.

https://github.com/NixOS/nix/issues/6656 here, in fact, is an empty of a regression in an error message! Sure it is not stable, but mere warnings should not say "error". That much we should promise.

Of course there is a cost to testing (regardless of the benefits), and so yes I don't want to promote a slippery slope where every bit of non-normative behavior still gets a test. That *would* be too much busywork. But as @fricklerhandwerk and others have argued, UI stuff like warnings and error messages is very visible to the user, and thus a very important influence on what they think of Nix.

It is for that reason that I say this bit of unstable behavior in particular deserves tests.

Ericson2314 commented:
My fancy command substitution had a race condition. I'll simply write to the two files and sed them afterwords, which avoids that. Sigh...

edolstra commented on tests/lang/parse-fail-uft8.err.exp:
```diff
@@ -0,0 +1,6 @@
+error: syntax error, unexpected invalid token, expecting end of file
```
This file should be named `parse-fail-utf8.err.exp`, not `parse-fail-uft8.err.exp`.

fricklerhandwerk commented on tests/lang.sh:
```diff
@@ -1,5 +1,46 @@
+
+# Flip which one is commented to update
+
+#doAccept=1
+doAccept=''
```
Maybe this should at least take an environment variable so it can be run slightly more conveniently?
Also add instructions to the hacking guide on how to deal with these tests in general, i.e.
- what to do when you change error messages?
- how to communicate these changes in PRs?

roberth replied:
Add instruction to update the `.err` to the error message when `.err` didn't match?
That makes the iteration cycle easier, and we can keep the hacking instructions simple as well.

fricklerhandwerk commented:
Triaged in Nix team meeting 2023-05-19:

- @edolstra: not keen on parsing `stderr`, although testing regressions error messages is worthwhile in principle
    - @ericson2314: we could have a helper script to regenerate all error messages
    - @roberth: the primary goal is to ensure that error messages keep working while we change the code around them.
- assigned to @roberth (with instructions to be very critical 😜)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-05-19-nix-team-meeting-minutes-56/28446/1

Ericson2314 commented:
I've added instructions on test failure and documentation.

I agree with @roberth the instructions should be mostly self-explanatory. But that gives the documentation the opportunity to focus on the "why" more than "how.

The hacking section has gotten very large, and so I moved the testing material to its own page.

fricklerhandwerk commented on doc/manual/src/contributing/testing.md:
```diff
@@ -0,0 +1,194 @@
+# Running tests
```
I encountered the same thing just yesterday. I'd quickly merge moving that section out into its own page if you split out a PR for that.

fricklerhandwerk commented on doc/manual/src/contributing/testing.md:
```diff
@@ -0,0 +1,194 @@
+
+### Characterization testing
+
+Occasionally, Nix utilizes a technique called [Characterization Testing](doc/manual/src/contributing/hacking.md) as part of the functional tests.
```
That link seems wrong

Ericson2314 commented:
Blocked on https://github.com/NixOS/nix/pull/8571

Ericson2314 commented:
OK, back to you, @roberth!

roberth commented on tests/lang/eval-fail-set.err.exp:
```diff
@@ -0,0 +1,7 @@
+error: undefined variable 'x'
```
Not great.

roberth commented on tests/lang.sh:
Did you run this through shellcheck by any chance?

roberth replied:
Hah you wrote that you did. :+1:

roberth commented on tests/lang.sh:
```diff
@@ -24,59 +60,89 @@ fail=0
+    echo 'to regenerate the files containing the expected output,'
+    echo 'and then view the git diff to decide whether a change is'
+    echo 'good/intenation or bad/accidental.'
+fi
```
If `_NIX_TEST_ACCEPT` is `1`, this hint should not be printed.
Instead, it could print whether any update happened, based on `$fail`. I guess the `1` case will be self-evident but that `0` case is not.

roberth replied:
I just found out that `$fail = 1` is not always indicative of a diff problem. It's also set when an exit code was wrong. It seems that we need a separate variable to track whether diffs were actually wrong. If it failed without a bad diff that means there's definitely a real issue, which can't be resolved with a `_NIX_TEST_ACCEPT=1`.

Instead of keeping the same meaning for `fail`, we might have an extra variable representing whether the run contains "definitely broken" cases because of the exit code.
At that point `fail` is derivable from the new variables, and we might as well check those instead of the redundant `fail` variable.

roberth commented on tests/lang.sh:
```diff
@@ -24,59 +60,89 @@ fail=0
+    echo ''
+    echo 'to regenerate the files containing the expected output,'
+    echo 'and then view the git diff to decide whether a change is'
+    echo 'good/intenation or bad/accidental.'
```
Typo

```suggestion
    echo 'good/intentional or bad/accidental.'
```

Not accidental as in complexity. Although that might be an uncommon interpretation, I don't think accidental adds much here.

```suggestion
    echo 'good/intentional or bad.'
```

```suggestion
    echo 'good/intentional or bad.'
    echo 'If the diff contains arbitrary or impure information,'
    echo 'please improve the normalization that the test applies to the output.'
```

More complete or too much hand holding?

roberth reviewed (commented):
I've read it through, but have yet to try it out.

Pre-existing problems can be scoped out with an issue.

Ericson2314 commented on tests/lang/eval-fail-hashfile-missing.err.exp:
```diff
@@ -0,0 +1,19 @@
+error:
+       … while calling the 'toString' builtin
```
On the main comments page, it *does* render correctly! :)

Ericson2314 replied:
It seems work tree encoding only makes sense if the file is something other than UTF-8? But the file is UTF-8 no conversion on checkout / check in should be necessary, right?

roberth replied:
I figured it might nudge github in the right direction, but your observation shows that it's likely just a bug on their end.

Ericson2314 commented on tests/lang/eval-fail-bad-antiquote-2.err.exp:
```diff
@@ -0,0 +1 @@
+error: operation 'addToStoreFromDump' is not supported by store 'dummy'
```
Ha! we're finding bugs already. Yay!

Ericson2314 commented:
> Pre-existing problems can be scoped out with an issue.

Are these 3 lingering issues the preexinsting ones you had in mind?

roberth reviewed (changes requested):
I've given this a try in #8607 and it seems to work well.
I'm a little concerned about the amount of untested code though. `lang.sh` now contains an untested test framework. Those functions should be split out, and we should have a test that checks that the framework behaves as expected.
Otherwise we might end with a situation worse than those already broken "addToStoreFromDump" test cases. Like ignored exit codes in some circumstance or some other weird undetected mistake.

roberth commented on tests/lang-test-infra.sh:
WIP? I'll draft the PR so you can request a review when done.

Ericson2314 replied:
Yes whoops, thanks

Ericson2314 commented:
OK, wrote the tests now!

fricklerhandwerk commented on doc/manual/src/contributing/testing.md:
```diff
@@ -86,6 +86,33 @@ GNU gdb (GDB) 12.1
+An interesting situation to document is the case when these tests are "overfitted".
+For example, the language tests store both the expected result in the case of success, and expected diagnostics (errors and warnings).
+The expected successful output is supposed to be highly stable --- we do not intend to make breaking changes to the (stable parts of) the Nix language.
+However, the errors and warnings are not stable in this way, we are free to change how they are displayed at any time.
```
```suggestion
However, the errors and warnings are not stable in this way.
We are free to change how they are displayed at any time.
```

fricklerhandwerk commented on doc/manual/src/contributing/testing.md:
```diff
@@ -86,6 +86,33 @@ GNU gdb (GDB) 12.1
+The expected successful output is supposed to be highly stable --- we do not intend to make breaking changes to the (stable parts of) the Nix language.
+However, the errors and warnings are not stable in this way, we are free to change how they are displayed at any time.
+
+Why bother to track these diagnostics if they are not stable then?
+The answer is that just because we Nix developers are free to change them whenever we like doesn't mean we don't care about them.
+On the contrary, errors and warnings are incredibly important to users.
+By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are code review.
+Regressions are caught, and improvements always show up in code review.
```
```suggestion
While diagnostic outputs are not a stable interface, they are important to users.
By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are under code review.
Regressions are caught, and improvements always show up in code review.
```

Ericson2314 replied:
I am worried that the new concise version rapidly goes over central point.

This rhetorical question was basically what @edolstra was asking, and I think it is a good question: anything that is not unstable is not normative behavior, so why the hell has we testing it? (emphasizing "not normative" is probably better than "not stable", I am realizing now.)

If you think the rhetorical question is too chatty, we can pick a different form, like

> It may be surprising that we would test non-normative behavior like diagnostic outputs.
> After all, the point of tests is usually to pin down just normative.
> If tests pinned down all behavior, normative and non-normative alike, they would just be a second copy of the implementation!

(but perhaps shorter!)

The point is not to write a question per se, but to get the readers attention --- whether the reader is already surprised or not. This is the key point, and we don't want them to miss it.

Ericson2314 replied:
On a meta point, while one might argue the main portion of the manual is strictly reference not explanation, I think the contributing chapter is free to cover all quadrants, as it currently does, as `nix.dev` etc. will not be covering contributing to Nix.

fricklerhandwerk replied:
> It may be surprising that we would test non-normative behavior like diagnostic outputs.

This one's good for prepending it to my suggestion. I'm in general against being chatty in documentation.

Ericson2314 replied:
```suggestion
It may be surprising that we would test non-normative behavior like diagnostic outputs.
While diagnostic outputs are not a stable interface, they are important to users.
By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are under code review.
Regressions are caught, and improvements always show up in code review.
```
OK!

fricklerhandwerk commented on doc/manual/src/contributing/testing.md:
```diff
@@ -86,6 +86,33 @@ GNU gdb (GDB) 12.1
+By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are code review.
+Regressions are caught, and improvements always show up in code review.
+
+These are good benefits, but its important that they are not offset by newly imposed costs.
+To that end, it much be very easy to regenerate the tests so characterization testing doesn't make it harder to intentionally change these unstable interfaces.
+That is why having an easy way to regenerate the expected output, e.g. with `_NIX_TEST_ACCEPT=1`, is very important.
```
```suggestion
To ensure that characterization testing doesn't make it harder to intentionally change these interfaces, there always must be an easy way to regenerate the expected output, e.g. with `_NIX_TEST_ACCEPT=1`.
```

roberth replied:
```suggestion
To ensure that characterization testing doesn't make it harder to intentionally change these interfaces, there always must be an easy way to regenerate the expected output, as we do with `_NIX_TEST_ACCEPT=1`.
```

fricklerhandwerk reviewed (commented):
Docs LGTM apart from comments.

Ericson2314 commented on tests/lang-test-infra.sh:
```diff
@@ -0,0 +1,72 @@
+# matches non-empty
+echo Hi! > "$TEST_ROOT/got"
+cp "$TEST_ROOT/got" "$TEST_ROOT/expected"
+badDiff=0
+diffAndAcceptInner test "$TEST_ROOT/got" "$TEST_ROOT/expected"
+(( "$badDiff" == 0 ))
```
```suggestion
(
  diffAndAcceptInner test "$TEST_ROOT/got" "$TEST_ROOT/expected"
  (( "$badDiff" == 0 ))
)
```

Ericson2314 commented on tests/lang-test-infra.sh:
```diff
@@ -0,0 +1,72 @@
+_NIX_TEST_ACCEPT=1 diffAndAcceptInner test "$TEST_ROOT/got" "$TEST_ROOT/expected"
+(( "$badDiff" == 0 ))
+
+# _NIX_TEST_ACCEPT doesn't match, non-empty `badDiff` set, file changed (was previously non-empty)
```
```suggestion
# _NIX_TEST_ACCEPT doesn't match, `badDiff=1` set, file changed (was previously non-empty)
```

Ericson2314 commented on doc/manual/src/contributing/testing.md:
```diff
@@ -86,6 +86,33 @@ GNU gdb (GDB) 12.1
+Occasionally, Nix utilizes a technique called [Characterization Testing](https://en.wikipedia.org/wiki/Characterization_test) as part of the functional tests.
+This technique is to include the exact output/behavior of a former version of Nix in a test in order to check that Nix continues to produce the same behavior going forward.
+
+For example, this technique is used for the language tests (`test/lang.sh`).
```
What to link? If we link any specific branch it could fall out of sync.

Mathnerd314 replied:
You could use a Github permalink, like https://github.com/NixOS/nix/blob/45a3e497da20e338b747c1e0ca1a1af5531080df/tests/lang.sh. Just make sure the same hash makes into the repo mainline.

roberth replied:
I think we should do the simple thing and fix the tooling if it doesn't render correctly, or not link at all. `lang.sh` *isn't* documentation, and most of the time you'll be _typing_ it in your editor / IDE / terminal anyway.

```suggestion
For example, this technique is used for the language tests ([`test/lang.sh`](../../../../test/lang.sh)).
```

Ericson2314 replied:
Yeah I want to document just one version, and if we can't make this go I rather just drop the reference to the file at all and just say "the language tests".

fricklerhandwerk replied:
Contributing stuff can point to GitHub master. Other than that I agree we should fix the tooling and make it render correctly by default.
